rfc/rfc-0014-binding-injection.md

```diff
@@ -0,0 +1,131 @@
+
+A duck type for exposing a binding is defined in RFC 0002. While not as wide spread, the Stream resource currently implements this duck type. Combining the Binding and PodSpecable duck types allows for a generic means to inject binding metadata and secrets into a consuming workload. A new CRD can express this injection:
+
+```yaml
+apiVersion: bindings.projectriff.io/v1alpha1
+kind: Injection
+metadata:
+  name: my-binding-injection
+spec:
+  binding:
+    apiVersion: streaming.projectriff.io/v1alpha1
+    kind: Stream
+    name: my-stream
+  containers:
+  - user-container
+  subject:
+    apiVersion: knative.projectriff.io/v1alpha1
+    kind: Deployer
+    name: my-deployer
+status:
+  observedGeneration: 1
+  conditions: []
+```
+
+- `.spec.binding` is an ObjectReference to a resource implementing the binding duck type
+- `.spec.containers` optionally restricts which containers are augmented, defaults to all containers
+- `.spec.subject` is an ObjectReference to a resource implementing the podspecable duck type
+- the subject and target much be in the same namespace as the injection resource
+
+The subject resource will be injected with the binding metadata and secret.
+
+```yaml
+apiVersion: streaming.projectriff.io/v1alpha1
+kind: Stream
+metadata:
+  name: my-stream
+spec: {}
+status:
+  binding:
+    metadataRef:
+      name: my-stream-binding-metadata
+    secretRef:
+      name: my-stream-binding-secret
+```
+
+```yaml
+apiVersion: knative.projectriff.io/v1alpha1
+kind: Deployer
+metadata:
+  name: before
+spec:
+  template:
+    spec:
+      containers:
+      - name: user-container
+        image: square
+```
+
+```yaml
+apiVersion: knative.projectriff.io/v1alpha1
+kind: Deployer
+metadata:
+  name: after
+spec:
+  template:
+    spec:
+      containers:
+      - name: user-container
+        image: square
+        env:
+        - name: CNB_BINDINGS
+          value: /var/riff/bindings
+        volumeMounts:
+        - mountPath: /var/riff/bindings/my-binding-injection/metadata
+          name: my-binding-injection-metadata
+          readOnly: true
+        - mountPath: /var/riff/bindings/my-binding-injection/secret
+          name: my-binding-injection-secret
+          readOnly: true
+    volumes:
+    - name: my-binding-injection-metadata
+      configMap:
+        name: my-stream-binding-metadata
+    - name: my-binding-injection-secret
+      secret:
+        secretName: my-stream-binding-secret
+```
+
+Notes:
+- if the `CNB_BINDINGS` env var is already defined, it must be respected
```

Comment:
This seems like maybe a mistake. CNB defines a default value for the environment variable already, so why not use that? Locating the bindings directory in `/var/riff` kind of blocks other options - you can't bind to something that lives in `/var/somethingelse` as well (because `CNB_BINDINGS` is global).

Reply:
Where does CNB define a default value for CNB_BINDINGS?

The situation you describe why this lines exists. CNB_BINDINGS can only hold a single value, if it is already defined, the new bindings should be mounted under that path.

Reply:
I guess we should clarify by "must be respected" means in practice then. I was under the impression that CNB_BINDINGS had a suggested default value. Maybe I misread their spec.

Reply:
boiling this down to the principles:
- we cannot change an existing CNB_BINDINGS var as there may be existing volume mounts located under that path
- the resulting container must have exactly one CNB_BINDINGS env var with the path which the volume mounts are mounted under

So, if there is an existing value, we need to add our volume mounts under the path that is defined by the existing var, otherwise we need to create the var with the path where we mounted the volumes.